### DISCUSSION TITLE
The validate-modules sanity test now also validates plugins

### DISCUSSION THREAD

**Discussion by felixfontein** [ansible-collections/news-for-maintainers#11] [Discussions on announcements]
Discussion for https://github.com/ansible-collections/news-for-maintainers/issues/9.

**Comment by briantist** [comment]
Thank you, is there a concise summary of what needs to change? Before & after kind of thing?

**Reply by briantist** [reply]
It looks like it's mainly changing doc strings in plugins to change `<plugin_type>: <plugin_name>` to `name: <plugin_name>`.

I can't tell if modules are also supposed to change from `module: <module_name>` to `name: <module_name>`, but it appears that's not the case since I'm only seeing failures on plugins.

**Reply by briantist** [reply]
I suppose `module: <module_name>` was also not used, at least not in executable code, because I just realized I have some test modules in integration targets that used `lookup: <module_name>` , I suppose if they were real content they might've failed some validation though.

**Reply by felixfontein** [reply]
I think `<plugin_type>: <plugin_name>` → `name: <plugin_name>` is the most common one. For modules, it stays to be `module: <module_name>` for now. (See also https://github.com/ansible/ansible/issues/71795, for plugins both forms were used all around ansible/ansible while everything still was in there.)

Another thing that will probably show up more often is specific to callback plugins: https://github.com/ansible/ansible/issues/71798 Instead of `callback_type: <type>` use `type: <type>`.

**Comment by evgeni** [comment]
I must admit I have a hard time to understand what `name` supposed to mean (and reading https://github.com/ansible/ansible/issues/71792 didn't really help, especially as it talks about it being optional, yet `validate-modules` raises `DOCUMENTATION.name: required key not provided @ data['name']. Got None` if unset).

My (previous) understanding was that `name` is the "loadable thing", so if I say "use the `theforeman.foreman.foreman` callback, the name is `theforeman.foreman.foreman`.

However, using the FQCN as the name raises `DOCUMENTATION.name: not a valid value for dictionary value @ data['name']. Got 'theforeman.foreman.foreman'` from `validate-modules`, using the non-FQCN name (hah) seems to work.

See https://github.com/theforeman/foreman-ansible-modules/pull/1354 for reference.

**Reply by felixfontein** [reply]
`name:` for plugins is the same as `module:` for modules. It's the 'short name', not a FQCN.

**Comment by gravesm** [comment]
It seems like non-module plugins don't allow the `complex` type in the RETURN block? This has been my experience with https://github.com/ansible-collections/kubernetes.core/blob/73499d9a0960835a3ed340286e7c7039d99fdc21/plugins/lookup/k8s.py#L158 at least. In this case, I've just replaced `contains` with `sample` which is probably a better solution anyways. Curious what the reason would be to not support `complex` in RETURN blocks, though?

**Reply by felixfontein** [reply]
I think you want to use `type: list`/`elements: dict` there, since you have a list of dictionaries.

**Reply by gravesm** [reply]
I tried that but still got errors on the fields in `contains` that used `type: complex`.

**Reply by sivel** [reply]
Looking at the code, `complex` is only valid for modules.  You would need to be explicit about the type in other plugin types.

**Comment by NilashishC** [comment]
With the new changes, usernames with a `/` (such as @ansible-collections/network) are not accepted by validate-modules. It should either be made a non-mandatory field or `/` should be allowed.

**Reply by KB-perByte** [reply]
+ 1 valid author should be a user or team, `/` is required for valid team input.

**Reply by felixfontein** [reply]
It was never accepted before (in modules), so this is not a new restriction. If you want such usernames accepted, please create a PR for ansible/ansible so that validate-modules accept them.

**Reply by NilashishC** [reply]
@felixfontein I understand that `/` wasn't accepted before, but usernames weren't mandatory before this either. Please correct me if I'm wrong.

**Reply by felixfontein** [reply]
@NilashishC the specific form for authors was mandatory before (for modules). Now the same code handles both modules and plugins.
